lib/internal/modules/esm/translators.js

```diff
@@ -86,32 +97,90 @@ translators.set('module', async function moduleStrategy(url) {
+    m.filename = filename;
+    m.paths = CJSModule._nodeModulePaths(m.path);
+    const resolved = CJSModule._resolveFilename(reexport, m);
+    if (resolved.endsWith('.json') || resolved.endsWith('.node'))
+      continue;
+    const reexportNames = parseCJSModuleExports(resolved);
+    for (const name of reexportNames)
+      exportNames.add(name);
+  }
+
+  return exportNames;
+}
+
 // Strategy for loading a node-style CommonJS module
 const isWindows = process.platform === 'win32';
 const winSepRegEx = /\//g;
-translators.set('commonjs', function commonjsStrategy(url, isMain) {
+translators.set('commonjs', async function commonjsStrategy(url, isMain) {
   debug(`Translating CJSModule ${url}`);
-  const pathname = internalURLModule.fileURLToPath(new URL(url));
-  const cached = this.cjsCache.get(url);
-  if (cached) {
-    this.cjsCache.delete(url);
-    return cached;
-  }
-  const module = CJSModule._cache[
-    isWindows ? StringPrototypeReplace(pathname, winSepRegEx, '\\') : pathname
-  ];
-  if (module && module.loaded) {
+
+  let filename = internalURLModule.fileURLToPath(new URL(url));
+  if (isWindows)
+    filename = StringPrototypeReplace(filename, winSepRegEx, '\\');
+
+  await ensureLexerInit();
+
+  const exportNames = parseCJSModuleExports(filename);
+  const namesWithDefault = exportNames.has('default') ?
+    [...exportNames] : ['default', ...exportNames];
+  return new ModuleWrap(url, undefined, namesWithDefault, function() {
+    debug(`Loading CJSModule ${url}`);
+    let module = CJSModule._cache[filename];
+    if (!module || !module.loaded) {
+      module = new CJSModule(filename);
+      module.filename = filename;
+      module.paths = CJSModule._nodeModulePaths(module.path);
+      CJSModule._cache[filename] = module;
+      module.load(filename);
+      assert(module && module.loaded);
+    }
+
     const exports = module.exports;
-    return new ModuleWrap(url, undefined, ['default'], function() {
+
+    let setDefault = false;
+    for (const exportName of exportNames) {
+      if (exportName === 'default') {
+        setDefault = exportNames.has('__esModule') &&
+          Boolean(exports.__esModule);
+        if (!setDefault)
+          continue;
+      }
+      // We might trigger a getter -> dont fail.
+      let value;
+      try {
+        value = exports[exportName];
+      } catch {}
+      this.setExport(exportName, value);
+    }
+
+    if (!setDefault)
       this.setExport('default', exports);
-    });
-  }
-  return new ModuleWrap(url, undefined, ['default'], function() {
-    debug(`Loading CJSModule ${url}`);
-    // We don't care about the return val of _load here because Module#load
-    // will handle it for us by checking the loader registry and filling the
-    // exports like above
-    CJSModule._load(pathname, undefined, isMain);
+
+    const cached = processLoader.ESMLoader.cjsCache.get(filename);
+    // GC the CJS source but retain exportNames info for future reexporter use.
```

Comment:
I feel like GC'ing the source could be done by the CJS loader since it knows better when it no longer needs the source?